solutions/usaco-599.mdx

```diff
@@ -31,127 +31,172 @@ a one-base polynomial hash to check if the substring appears only once. If we
+signed main() {
+	freopen("lightsout.in", "r", stdin);
+	freopen("lightsout.out", "w", stdout);
+
+	cin.tie(0)->sync_with_stdio(0);
+
+	int n;
+	cin >> n;
+
+	vector<coord> dt(n);
+	for (int i = 0; i < n; i++) {
+		cin >> dt[i].x >> dt[i].y;
 	}
-}
 
-int cwdist[205][205], dist[205];
-void gen_dist()
-{
-	for(int i = 0; i < n; i++)
-	{
-		for(int j = i; j < n; j++)
-		{
-			cwdist[i][(j+1)%n] = cwdist[i][j] + abs((arr[j].f - arr[(j+1)%n].f) + (arr[j].s - arr[(j+1)%n].s));
-			if(j+1 != n) cwdist[(j+1)%n][i] = INT_MAX;
+	function<int(int i, int j)> calc_dist = [&](int i, int j) {
+		return abs(dt[i].x - dt[j].x) + abs(dt[i].y - dt[j].y);
+	};
+
+	vector<vector<vector<int>>> dst(n,
+																	vector<vector<int>>(n,
+																											vector<int>(2, INT32_MAX)));
+
+	// i, j, 0 - clockwise dist from i -> j
+	// i, j, 1 - counter-clockwise dist from i -> j
+	for (int i = 0; i < n; i++) {
+		for (int j = i; j < i + n; j++) {
+			if (i == j) {
+				dst[i][j][0] = 0;
+			} else {
+				int x = (j + n) % n;
+				int y = (j - 1 + n) % n;
+				dst[i][x][0] = dst[i][y][0] + calc_dist(x, y);
+			}
 		}
 	}
-	for(int i = 0; i < n; i++)
-	{
-		dist[i] = min(cwdist[0][i], cwdist[i][0]);
+	for (int i = 0; i < n; i++) {
+		for (int j = i; j > i - n; j--) {
+			if (i == j) {
+				dst[i][j][1] = 0;
+			} else {
+				int x = (j + n) % n;
+				int y = (j + 1 + n) % n;
+				dst[i][x][1] = dst[i][y][1] + calc_dist(x, y);
+			}
+		}
 	}
-}
 
-int main()
-{
-	freopen("lightsout.in", "r", stdin);
-	freopen("lightsout.out", "w", stdout);
-	cin >> n;
-	for(int i = 0; i < n; i++)
-	{
-		cin >> arr[i].f >> arr[i].s;
-	}
-	k.push_back(0);
-	for(int i = 1; i < n; i++)
-	{
-		int pre = i-1;
-		k.push_back(abs(arr[pre].f-arr[i].f)+abs(arr[pre].s-arr[i].s));
-		int nxt = (i+1)%n;
-		if(arr[i].f - arr[pre].f > 0)
-		{
-			if(arr[nxt].s - arr[i].s > 0) k.push_back(1100001);
-			else k.push_back(1100002);
-		} else if(arr[i].f - arr[pre].f < 0)
-		{
-			if(arr[nxt].s - arr[i].s > 0) k.push_back(1100002);
-			else k.push_back(1100001);
-		} else
-		{
-			if(arr[i].s - arr[pre].s > 0)
-			{
-				if(arr[nxt].f - arr[i].f > 0) k.push_back(1100002);
-				else k.push_back(1100001);
-			} else if(arr[i].s - arr[pre].s < 0)
-			{
-				if(arr[nxt].f - arr[i].f > 0) k.push_back(1100001);
-				else k.push_back(1100002);
-			}
+	vector<int> ang(n);
+	for (int i = 0; i < n; i++) {
+		int prev = (i - 1 + n) % n;
+		int nxt = (i + 1) % n;
+
+		/*
+		 * two type of possible angles (90 deg either on inside or outside)
+		 * reminder: all edges are parallel to either x or y axis & alternate
+		 *
+		 * |--------------------------|
+		 * | ^                     ^  |
+		 * |      angles of 90°       |
+		 * | ˅                     ˅  |
+		 * |--------------------------|
+		 *
+		 * |-------------------------|
+		 * |     angles of 270°      |
+		 * |              > ----------
+		 * |                |
+		 * |                |
+		 * |-----------------
+		 */
```

Comment:
i'd just put this in the editorial somewhere

Reply:
ill draw it up soon